### DISCUSSION TITLE
Ability to specify upper time limit spent per repository

### DISCUSSION THREAD

**Discussion by MShekow** [renovatebot/renovate#14276] [Suggest an Idea]
I am operating Renovate Bot on a self-hosted GitLab instance. It usually works fine, using an hourly CI scheduled pipeline. Renovate Bot usually spends about 5-40 seconds per repo. However, sometimes, especially for Java/Maven-based repos with unreachable Maven repos, it can take up to 37 minutes (for a _single_ repo). This completely breaks my Renovate Bot pipeline, where Renovate also has to visits hundreds of other repos, for which it needs also about 30 minutes.

Is there any way I can configure Renovate Bot to abort visiting a repo if it has not completed its work during X minutes (and then maybe have the bot create an issue about the slowness in the repo)? Renovate should then not abort completely, but continue with the other (auto-discovered) repos.

FYI, the "slowness" is due to the fact that there are many log entries like this one (simplified):
```
2022-02-03T10:37:14.118Z [project/path] Looking up some:other-dependency in repository https://unreachable.host.com/somepath/
2022-02-03T10:38:51.393Z [project/path] Cannot connect to maven host
```

**Comment by rarkins** [comment]
Due to node's asynchronous nature, this wouldn't be easy. I think we'd be better off reproducing this type of unreachable host and building in:
(a) lower default timeouts, and
(b) "remembering" if a certain host is unreachable and skipping it immediately for subsequent lookups

**Reply by MShekow** [reply]
Sounds good to me. Do you have concrete plans, or is there a way I can help with diagnosing something?

**Reply by rarkins** [reply]
The best thing would be if you can create a public reproduction repo which demonstrates the "long timeouts" problem. It doesn't necessarily need to be with your own hosts, but it must somehow reproduce the same result so that we can debug it and work out how to shortcut such timeouts. If you can do that, please create a feature request linking to the reproduction. The request should be "Detect and avoid repeated queries to unreachable hosts". Thanks

**Reply by MShekow** [reply]
Thanks. As it turns out, the problem is caused by the specific Maven repository, which is operated by our company, but I have not given the permission to disclose it. It is configured incorrectly somehow, s.t. any client that attempts to establish a SSL connection (HTTPS) to it will fail after a while (~ 30 seconds), with a "connection closed" error. If I simply use a different (non-existent / non-resolvable) host, the problem does not occur, and the job completes after about 80 seconds (instead of 30 minutes).
